src/complex_fluids/CFGiesekusRelaxation.cpp

```diff
@@ -0,0 +1,153 @@
+        }
+        else if (d_log_conform)
+        {
+            Eigen::Matrix2d mat;
+            mat(0, 0) = (*in_data)(idx, 0);
+            mat(1, 1) = (*in_data)(idx, 1);
+            mat(0, 1) = mat(1, 0) = (*in_data)(idx, 2);
+            mat = mat.exp();
+            Q_xx = mat(0, 0);
+            Q_yy = mat(1, 1);
+            Q_xy = mat(0, 1);
+        }
+        else
+        {
+            Q_xx = (*in_data)(idx, 0);
+            Q_yy = (*in_data)(idx, 1);
+            Q_xy = (*in_data)(idx, 2);
+        }
+        (*ret_data)(idx, 0) =
+            l_inv * (-1.0 * (d_alpha * (Q_xx * Q_xx + Q_xy * Q_xy) + (1.0 - 2.0 * d_alpha) * Q_xx + (d_alpha - 1.0)));
+        (*ret_data)(idx, 1) =
+            l_inv * (-1.0 * (d_alpha * (Q_yy * Q_yy + Q_xy * Q_xy) + (1.0 - 2.0 * d_alpha) * Q_yy + (d_alpha - 1.0)));
+        (*ret_data)(idx, 2) = l_inv * (-1.0 * (d_alpha * (Q_xx * Q_xy + Q_xy * Q_yy) + (1.0 - 2.0 * d_alpha) * Q_xy));
+#endif
+#if (NDIM == 3)
+        double Q_xx, Q_yy, Q_zz, Q_yz, Q_xz, Q_xy;
+        if (d_sqr_root)
+        {
+            Q_xx = (*in_data)(idx, 0) * (*in_data)(idx, 0) + (*in_data)(idx, 5) * (*in_data)(idx, 5) +
+                   (*in_data)(idx, 4) * (*in_data)(idx, 4);
+            Q_yy = (*in_data)(idx, 5) * (*in_data)(idx, 5) + (*in_data)(idx, 1) * (*in_data)(idx, 1) +
+                   (*in_data)(idx, 3) * (*in_data)(idx, 3);
+            Q_zz = (*in_data)(idx, 4) * (*in_data)(idx, 4) + (*in_data)(idx, 3) * (*in_data)(idx, 3) +
+                   (*in_data)(idx, 2) * (*in_data)(idx, 2);
+            Q_yz = (*in_data)(idx, 5) * (*in_data)(idx, 4) + (*in_data)(idx, 1) * (*in_data)(idx, 3) +
+                   (*in_data)(idx, 3) * (*in_data)(idx, 2);
+            Q_xz = (*in_data)(idx, 0) * (*in_data)(idx, 4) + (*in_data)(idx, 5) * (*in_data)(idx, 3) +
+                   (*in_data)(idx, 4) * (*in_data)(idx, 2);
+            Q_xy = (*in_data)(idx, 0) * (*in_data)(idx, 5) + (*in_data)(idx, 5) * (*in_data)(idx, 1) +
+                   (*in_data)(idx, 4) * (*in_data)(idx, 3);
+        }
+        else if (d_log_conform)
+        {
+            Eigen::Matrix3d mat;
+            mat(0, 0) = (*in_data)(idx, 0);
+            mat(1, 1) = (*in_data)(idx, 1);
+            mat(2, 2) = (*in_data)(idx, 2);
+            mat(1, 2) = mat(2, 1) = (*in_data)(idx, 3);
+            mat(0, 2) = mat(2, 0) = (*in_data)(idx, 4);
+            mat(0, 1) = mat(1, 0) = (*in_data)(idx, 5);
+            mat = mat.exp();
+            Q_xx = mat(0, 0);
+            Q_yy = mat(1, 1);
+            Q_zz = mat(2, 2);
+            Q_xy = mat(0, 1);
+            Q_xz = mat(0, 2);
+            Q_yz = mat(1, 2);
+        }
+        else
+        {
+            Q_xx = (*in_data)(idx, 0);
+            Q_yy = (*in_data)(idx, 1);
+            Q_zz = (*in_data)(idx, 2);
+            Q_yz = (*in_data)(idx, 3);
+            Q_xz = (*in_data)(idx, 4);
+            Q_xy = (*in_data)(idx, 5);
+        }
+        (*ret_data)(idx, 0) =
```

Comment:
This last bit looks like the only part that is specific to Giesekus. I wonder if it would be better to template the base class on the function that evaluates these kinds of expressions, or to have users pass in these kinds of functions as simple C functions (similar to the stress and force functions in `IBFEMethod`). Presumably would make it easier for users to specify their own models.